### PR TITLE
[Enhancement](hudi) Support hudi mor only java side ,be side  not support

### LINKED ISSUE #18432: [Feature] (Hudi) Support Hudi MOR table
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

Support reading Hudi MOR table by using [JNI Connector](https://github.com/apache/doris/pull/17960)

### Use case

_No response_

### Related issues

You can refer to this PR #17960 and following classes:

1. be/src/vec/exec/jni_connector.h
2. be/src/vec/exec/scan/jni_reader.h
3. fe/java-udf/src/main/java/org/apache/doris/jni/MockJniScanner.java

Maybe you can use [Presto-Hudi-Bundle](https://mvnrepository.com/artifact/org.apache.hudi/hudi-presto-bundle) to implement this.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by lexluo09**
# Proposed changes

Issue Number: #18432

## Problem summary

Support reading Hudi MOR table by using jni connector. 
Note:
  the FE part of the current PR is not completed all, and the BE part will be supplemented in next PR.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/catalog/external/HudiExternalTable.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+                    getName(), dbName);
+            tTableDescriptor.setHudiTable(hudiTable);
+            return tTableDescriptor;
+        }
+        throw new UnsupportedOperationException("hudi not supported  the catalog type :" + hudiCatalogType);
+    }
+
+    public static String fromAvroHudiTypeToHiveTypeString(Schema avroSchema) {
+        Schema.Type columnType = avroSchema.getType();
+        switch (columnType) {
```
Comment (AshinGau) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/external/HudiExternalTable.java:76]: How about `decimal` and `timestamp` ... types?

File: fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java [hunk 2]
```diff
@@ -2431,14 +2432,17 @@ private void createHudiTable(Database db, CreateTableStmt stmt) throws DdlExcept
         HiveConf hiveConf = new HiveConf();
         hiveConf.set(HMSProperties.HIVE_METASTORE_URIS,
                 hudiTable.getTableProperties().get(HudiProperty.HUDI_HIVE_METASTORE_URIS));
+
```
Comment (AshinGau) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java:2435]: Many unexpected changes.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/external/HudiExternalDatabase.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import org.apache.doris.persist.gson.GsonPostProcessable;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class HudiExternalDatabase extends ExternalDatabase<HudiExternalTable> implements GsonPostProcessable {
```
Comment (AshinGau) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/external/HudiExternalDatabase.java:33]: We should register the subtype of `DatabaseIf` in `org.apache.doris.persist.gson.GsonUtils`

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiScanNode.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+        List<FieldSchema> allFields = targetTable.getRemoteTable().getSd().getCols();
+        allFields.addAll(targetTable.getRemoteTable().getPartitionKeys());
+
+        for (Schema.Field hudiField : hudiSchema.getFields()) {
+
+            String columnName = hudiField.name().toLowerCase(Locale.ROOT);
+            Optional<FieldSchema> field = allFields.stream()
+                    .filter(f -> f.getName().equals(columnName)).findFirst();
+            if (!field.isPresent()) {
+                String errorMsg = String.format("Hudi column %s not exists in hive metastore.", hudiField.name());
```
Comment (AshinGau) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiScanNode.java:168]: How is this situation caused? Can the [synchronization tools](https://hudi.apache.org/docs/syncing_metastore#hive-sync-tool) maintain strong consistency in the table schema of the hive schema?

File: fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiScanNode.java [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+            columnTypes.add(columnType);
+        }
+
+        try {
+            String globPath = String.format("%s/%s/*", hudiClient.getBasePath(), partitionName);
+            List<FileStatus> statuses = FSUtils.getGlobStatusExcludingMetaFolder(hudiClient.getRawFs(),
+                    new Path(globPath));
+
+            HoodieTimeline timeline = hudiClient.getCommitsAndCompactionTimeline().filterCompletedInstants();
+            Option<HoodieInstant> latestInstant = timeline.lastInstant();
```
Comment (AshinGau) [inline fe/fe-core/src/main/java/org/apache/doris/planner/external/hudi/HudiScanNode.java:182]: So currently only support MOR from the last snapshots. Can we reserve interfaces to support `Time Travel Query` and `Incremental query`.

**Comment by AshinGau** [comment]
Cloud you please add in the description that the FE part of the current PR is completed, and the BE part will be supplemented in next PR.

**Comment by AshinGau** [comment]
LGTM

**Comment by morningman** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.91 seconds
 stream load tsv:          420 seconds loaded 74807831229 Bytes, about 169 MB/s
 stream load json:         23 seconds loaded 2358488459 Bytes, about 97 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          78.0 seconds inserted 10000000 Rows, about 128K ops/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230524091604_clickbench_pr_149641.html

**Comment by DongLiang-0** [comment]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**